src/pure/inmost.ml

```diff
@@ -11,41 +11,36 @@ include Status
 type bigstring = Body.bigstring
 
 type upload_event = [
-  | `File of string * string
+  | `Part of string option * string option
   | `Field of string * string
   | `Done
   | `Wrong_content_type
 ]
 
 (* Used for converting the push interface of Multipart_form_data into the pull
    interface of Dream. *)
-type multipart_state = {
-  mutable initial : bool;
-  mutable event_listener : upload_event Lwt.u option;
-  mutable chunk_listener : string option Lwt.u option;
-  mutable last_field_name : string option;
-  mutable last_file_name : string option;
-  mutable buffered_chunk : string option;
-  mutable next_file : bool;
-  mutable continue : unit Lwt.u;
-  mutable fields : bool;
-  mutable field : unit -> (string * string) option Lwt.t;
-}
+type uid = < >
```

Comment:
`Multipart_form` asks to "identify" a part because it considers that a stream of a part can be consumed concurrently with the parsing of the whole body. The idea is to use this unique identifier to identify a part and to bind it with a `Leaf` of the returned `Multipart_form.t`.

With such a bind, we are able to retrieve the stream of the part from the unique ID and the part (including its header) from the _tree_ `Multipart_form.t`. Generally, this unique ID can be an integer or something else which derives from the given `Header.t`. This how `multipart_form` was designed.

In this specific context, the `uid` is used to differentiate 2 states when we process the `multipart/form-data`:
- when we start the process (`uid = Dream.epsilon`)
- when we manipulate a part

In a certain sense, `uid` is like the old `state.initial` value used before this PR. If you want to rollback to such value, I can propose something 👍 .

Reply:
> that a stream of a part can be consumed concurrently with the parsing of the whole body.

Does this mean something like, you are effectively parsing a node of a multipart tree, while directly parsing one of its leaves? Because otherwise I don't see how *one* body can be consumed concurrently (with itself?), since the parts and their whole structure arrives serialized.

Can the uid be just an integer? My guess is that the uid has to be unique within one message body. If that's right, I think, even on a 32-bit system, we are unlikely to have so many concurrent multipart parsers, that an id could get incremented 2^31 times and wrap all the way around to a collision within a single stream. Of course, that leaves the question of what to use for epsilon. I hadn't even realized that such a value would be necessary. Maybe it's possible to increment by 2 and use an odd number for epsilon :D Then we have 2^30 ids margin for collision.

> In a certain sense, `uid` is like the old `state.initial` value used before this PR. If you want to rollback to such value, I can propose something 👍 .

It's not necessary to roll back. I just don't see the motivation for specifically this kind of API, yet :P However, I will accept it anyway, but I want to see if there is something less mysterious to future readers (including me) that objects :) And hopefully doesn't require conditional code.

Reply:
I deleted the usage of `type uid = < >` and choose to have something more _readable_ for a maintainer. A state still is needed to separate the initial operation (launch the consumer stream), the header state to emit `Field` and the part state to emit contents. I added an explanation about what is going on 👍 .